plugin.py

```diff
@@ -49,26 +49,50 @@ def prepare(self):
         if os.path.isfile(self.module_output):
             os.remove(self.module_output)
 
-    # run inside gui-thread-safe context
-    def invoke(self, runner):
-        wx.CallAfter(runner)
-
+    # export board.dsn file from pcbnew
     def RunExport(self):
-        pcbnew.ExportSpecctraDSN(self.module_input)
+        ok = pcbnew.ExportSpecctraDSN(self.module_input)
+        if ok and os.path.isfile(self.module_input):
+            return True
+        else:
+            show_error("Failed to invoke pcbnew.ExportSpecctraDSN")
+            return False
         
+    # auto route by invoking FreeRouting.jar
     def RunRouter(self):
         command = [self.java_path, "-jar", self.module_path, "-de", self.module_input, "-s"]
-        process = subprocess.Popen(command)
-        process.wait(self.module_timeout)
+        try:
+            process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            output, error = process.communicate()
+            if process.returncode == 0:
+                return True
+            else:  
+                show_error("Java program failure: %d\n %s\n %s\n" % (process.returncode, output, error))
+                return False
+        except Exception as error:
+                show_error("Java process failure: %s" % str(error))
+                return False
 
+    # import generated board.ses file into pcbnew
     def RunImport(self):
-        pcbnew.ImportSpecctraSES(self.module_output)
+        ok = pcbnew.ImportSpecctraSES(self.module_output)
+        if ok and os.path.isfile(self.module_output):
+            return True
+        else:
+            show_error("Failed to invoke pcbnew.ImportSpecctraSES")
+            return False
+        
+    # invoke chain of dependent methods
+    def RunSteps(self):
+        return self.RunExport() and self.RunRouter() and self.RunImport()
 
+    # kicad plugin action entry
     def Run(self):
+        if not has_pcbnew_api():
+            show_error("Missing required python API:\n pcbnew.ExportSpecctraDSN\n pcbnew.ImportSpecctraSES\n")
+            return
         self.prepare()
-        self.invoke(self.RunExport)
-        self.invoke(self.RunRouter)
-        self.invoke(self.RunImport)
+        safe_invoke(self.RunSteps)
```

Comment:
It's not good to wait on java in wx gui thread. Limit what you run in safe_invoke to only what is necessary, in this case import.

Reply:
the intent is:
1. block any `pcbnew` state changes while `java` is active
2. invoke `self.RunSteps` as single unit of dependent chained methods

since `safe_invoke` is asynchronous, how would you do that?

Reply:
You can't do it properly with current API we have. 
Problem with blocking UI thread is that OS window manager events will not be processed either so you will get "Not responding" status with potentially more side effects.
At this point I think it's ok to leave it up to user: if you change the pcb while autorouting happens and then import breaks your changes it's on you.
Not running java in UI thread will also allow you to have a dialog with a cancel button if user changes his mind and doesn't want to wait for autorouter or import the result.

Reply:
please review a hack around this problem